<a href="https://colab.research.google.com/github/yqian000/csc448AI/blob/main/a4_Yue.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Imports and pip installations (if needed)
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

import pandas as pd
import numpy as np


# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
data = load_iris()
df = pd.DataFrame(data=data.data, columns=data.feature_names)
df["target"] = data.target

# Output the first 15 rows of the data
df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  
10       0  
11       0  
12       0  
13       0  
14       0

In [3]:
# Display a summary of the table information (number of datapoints, etc.)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


In [4]:
df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  \
count         150.000000        150.000000         150.000000   
mean            5.843333          3.057333           3.758000   
std             0.828066          0.435866           1.765298   
min             4.300000          2.000000           1.000000   
25%             5.100000          2.800000           1.600000   
50%             5.800000          3.000000           4.350000   
75%             6.400000          3.300000           5.100000   
max             7.900000          4.400000           6.900000   

       petal width (cm)      target  
count        150.000000  150.000000  
mean           1.199333    1.000000  
std            0.762238    0.819232  
min            0.100000    0.000000  
25%            0.300000    0.000000  
50%            1.300000    1.000000  
75%            1.800000    2.000000  
max            2.500000    2.000000

## About the dataset
The dataset as 150 instances in total (50 in each of three classes). There is no missing attributes in the dataset.  

The features contain 4 numeric attributes described as follows:  
* sepal length in cm  
* sepal width in cm  
* petal length in cm  
* petal width in cm   

The label is the target column which describes the class of the speices, the label maps the actual class to a numeric value as follows:  
* class 1: Iris-Setosa represented as 0
* class 2: Iris-Versicolour represented as 1
* class 3: Iris-Virginica represented as 2  


# Part 2: Split the dataset into train and test

In [5]:
# Take the dataset and split it into our features (X) and label (y)
X = df[['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']]
y = df['target']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# validate
print(f"X_train shape is {X_train.shape}")
print(f"X_test shape is {X_test.shape}")

X_train shape is (135, 4)
X_test shape is (15, 4)


# Part 3: Logistic Regression

In [6]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression(max_iter=200).fit(X_train, y_train)

In [7]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_x = np.array([[5, 3, 4, 1.1]])
sample_y = model.predict_proba(sample_x)[0]
print(f"The probability of being class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of being class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of being class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of being class Iris-Setosa (0) is: 0.06472 
The probability of being class Iris-Versicolour (1) is: 0.90480 
The probability of being class Iris-Virginica (2) is: 0.03048 


/Users/yueqian/opt/anaconda3/envs/appliedml/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


In [8]:
# iii. Report on the score for Logistic regression model, what does the score measure?
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)
print(f"Train score is {train_score:.5f} and Test Score is {test_score:.5f}.")

Train score is 0.97778 and Test Score is 1.00000.


The score measures the accuracy of the model.  
The train score is very close to positive 1 which means that the model has a high accuaracy score and predicts well on most of the labels.  
The test score is 1 meaning the model makes perfect prediction everytime.

In [9]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
print(f"The coefficents are \n{model.coef_}")
print(f"The intercepts are \n{model.intercept_}")

The coefficents are 
[[-0.42671896  0.97265658 -2.44465216 -1.03186382]
 [ 0.51310009 -0.22373863 -0.21510023 -0.85159643]
 [-0.08638113 -0.74891795  2.65975239  1.88346025]]
The intercepts are 
[  9.50104298   1.91225076 -11.41329374]


# Part 4: Support Vector Machine

In [10]:
# i. Use sklearn to train a Support Vector Classifier on the training set
model = SVC(probability=True, random_state=42).fit(X_train, y_train)

In [11]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_x = np.array([[5, 3, 4, 1.1]])
sample_y = model.predict_proba(sample_x)[0]
print(f"The probability of being class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of being class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of being class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of being class Iris-Setosa (0) is: 0.01962 
The probability of being class Iris-Versicolour (1) is: 0.97147 
The probability of being class Iris-Virginica (2) is: 0.00891 


/Users/yueqian/opt/anaconda3/envs/appliedml/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


In [12]:
# iii. Report on the score for the SVM, what does the score measure?
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)
print(f"Train score is {train_score:.5f} and Test Score is {test_score:.5f}.")

Train score is 0.97037 and Test Score is 1.00000.


The score measures the accuracy of the model.  
The train score is very close to positive 1 which means that the model has a high accuaracy score and predicts well on most of the labels.  
The test score is 1 meaning the model makes perfect prediction everytime.

# Part 5: Neural Network

In [13]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
model = MLPClassifier(max_iter=600, random_state=42).fit(X_train, y_train)

In [14]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_x = np.array([[5, 3, 4, 1.1]])
sample_y = model.predict_proba(sample_x)[0]
print(f"The probability of being class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of being class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of being class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of being class Iris-Setosa (0) is: 0.00552 
The probability of being class Iris-Versicolour (1) is: 0.98981 
The probability of being class Iris-Virginica (2) is: 0.00467 


/Users/yueqian/opt/anaconda3/envs/appliedml/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  warnings.warn(


In [15]:
# iii. Report on the score for the Neural Network, what does the score measure?
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)
print(f"Train score is {train_score:.5f} and Test Score is {test_score:.5f}.")

Train score is 0.98519 and Test Score is 1.00000.


The score measures the accuracy of the model.  
The train score is very close to positive 1 which means that the model has a high accuaracy score and predicts well on most of the labels.  
The test score is 1 meaning the model makes perfect prediction everytime.

### iv: Experiment with different options for the neural network  
The best score: `Train score = 0.99 and Test Score =1`.  The model that achieved this score has `max_iter=600` so that it converges and everything else is left at default.

# Part 6: K-Nearest Neighbors

In [16]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
model = KNeighborsClassifier().fit(X_train, y_train)

In [17]:
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_x = np.array([[5, 3, 4, 1.1]])
sample_y = model.predict_proba(sample_x)[0]
print(f"The probability of being class Iris-Setosa (0) is: {sample_y[0]:.5f} ")
print(f"The probability of being class Iris-Versicolour (1) is: {sample_y[1]:.5f} ")
print(f"The probability of being class Iris-Virginica (2) is: {sample_y[2]:.5f} ")

The probability of being class Iris-Setosa (0) is: 0.00000 
The probability of being class Iris-Versicolour (1) is: 1.00000 
The probability of being class Iris-Virginica (2) is: 0.00000 


/Users/yueqian/opt/anaconda3/envs/appliedml/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


In [18]:
# iii. Report on the score for kNN, what does the score measure?
train_score = model.score(X_train, y_train)
test_score = model.score(X_test, y_test)
print(f"Train score is {train_score:.5f} and Test Score is {test_score:.5f}.")

Train score is 0.97037 and Test Score is 1.00000.


The score measures the accuracy of the model.  
The train score is very close to positive 1 which means that the model has a high accuaracy score and predicts well on most of the labels.  
The test score is 1 meaning the model makes perfect prediction everytime.

# Part 7: Conclusions and takeaways

All the models turned out to have a pretty good performace on the dataset. One of the reasons could be that the features have high correlation with the label which helped the models to perform better.  

The Netural Network classifier performed the best on the dataset since it had the highest train and test score. One of the reasons could be the Netural Network classifier produced a more complex model instead of a simple linear model which lead to a better classifer.  

I was surprised that the test scores are 1 for all models. This might be that the test set is very small and all predicted labels turned out to match the true labels. 